drivers/net/phy/marvell10g_ptp.c

```diff
@@ -0,0 +1,453 @@
+	if (!priv)
+		return NULL;
+
+	priv->phydev = phydev;
+	priv->extts_enabled = false;
+	spin_lock_init(&priv->lock);
+
+	priv->caps.owner = THIS_MODULE;
+	strscpy(priv->caps.name, "mv10g-phy-phc", sizeof(priv->caps.name));
+	priv->caps.max_adj = 0;
+	priv->caps.n_alarm = 0;
+	priv->caps.n_ext_ts = 1;
+	priv->caps.n_per_out = 0;
+	priv->caps.n_pins = 0;
+	priv->caps.pps = 0;
+	priv->caps.pin_config = NULL;
+	priv->caps.adjfine = mv3310_adjfine;
+	priv->caps.adjphase = mv3310_adjphase;
+	priv->caps.adjtime = mv3310_adjtime;
+	priv->caps.gettimex64 = mv3310_gettimex64;
+	priv->caps.settime64 = mv3310_settime64;
+	priv->caps.enable = mv3310_enable;
+	priv->caps.verify = mv3310_verify;
+	priv->caps.do_aux_work = mv3310_do_aux_work;
+	/* This is set to NULL instead of EOPNOTSUPP, simply defining it will
+	   present "has cross timestamping support" in capabilities. */
+	priv->caps.getcrosststamp = NULL;
+
+	priv->clock = ptp_clock_register(&priv->caps, &phydev->mdio.dev);
+	if (IS_ERR(priv->clock)) {
+		dev_err(&phydev->mdio.dev, "failed to register PTP clock\n");
+		devm_kfree(&phydev->mdio.dev, priv);
+		return NULL;
+	}
+
+	return priv;
+}
+
+int mv3310_ptp_power_up(struct phy_device *phydev)
+{
+	if (!mv3310_is_ptp_supported(phydev))
+		return 0;
+
+	/* Enable M unit used for PTP */
+	return phy_set_bits_mmd(phydev, MDIO_MMD_VEND2, MV_V2_MODE_CFG,
+				MV_V2_MODE_CFG_M_UNIT_PWRUP);
+}
+
+int mv3310_ptp_power_down(struct phy_device *phydev)
+{
+	if (!mv3310_is_ptp_supported(phydev))
+		return 0;
+
+	return phy_clear_bits_mmd(phydev, MDIO_MMD_VEND2, MV_V2_MODE_CFG,
+				  MV_V2_MODE_CFG_M_UNIT_PWRUP);
+}
+
+static int mv3310_adjfine(struct ptp_clock_info *ptp, long scaled_ppm)
+{
+	return -EOPNOTSUPP;
+}
+
+static int mv3310_adjphase(struct ptp_clock_info *ptp, s32 phase)
+{
+	return -EOPNOTSUPP;
+}
+
+static int mv3310_verify(struct ptp_clock_info *ptp, unsigned int pin,
+			 enum ptp_pin_function func, unsigned int chan)
+{
+	return -EOPNOTSUPP;
+}
+
+static int mv3310_read_ptp_reg(struct phy_device *phydev, u32 regnum,
+			       u32 *regval)
+{
+	int ret;
+
+	/* Read register address */
+	ret = phy_read_mmd(phydev, MDIO_MMD_VEND2, regnum);
+	if (ret < 0)
+		return ret;
+
+	/* Read that Indirect_read_address gives requested address */
+	ret = phy_read_mmd(phydev, MDIO_MMD_VEND2, MV_V2_INDIRECT_READ_ADDR);
+	if (ret < 0)
+		return ret;
+	if (ret != regnum) {
+		pr_err("Indirect read address mismatch: %04x != %04x\n", ret,
+		       regnum);
+		return -EINVAL;
+	}
+
+	/* Read Indirect_read_data_low provides lower 16-bits (15:0) of data */
+	ret = phy_read_mmd(phydev, MDIO_MMD_VEND2,
+			   MV_V2_INDIRECT_READ_DATA_LOW);
+	if (ret < 0)
+		return ret;
+	*regval = ret & 0xffff;
+
+	/* Read Indirect_read_data_high provides upper 16-bits (31:16) of data */
+	ret = phy_read_mmd(phydev, MDIO_MMD_VEND2,
+			   MV_V2_INDIRECT_READ_DATA_HIGH);
+	if (ret < 0)
+		return ret;
+	*regval += ((ret & 0xffff) << 16);
```

Comment:
isn't this is the right operation to ensure low bits aren't affected
```suggestion
	*regval |= ((ret & 0xffff) << 16);
```

Reply:
at this point regval has only the LS 16 bits set, so += and |= will do the same thing. I know it's best to do bitwise ops on unsigned values, but here ret is checked to be positive so no need for cast